src/core/qgspointlocator.cpp

```diff
@@ -471,6 +568,100 @@ class QgsPointLocator_VisitorVerticesInRect : public IVisitor
     QgsPointLocator::MatchFilter *mFilter = nullptr;
 };
 
+/**
+ * \ingroup core
+ * Helper class used when traversing the index looking for centroid - builds a list of matches.
+ * \note not available in Python bindings
+ * \since QGIS 3.10
+*/
+class QgsPointLocator_VisitorCentroidsInRect : public IVisitor
+{
+  public:
+    //! Constructs the visitor
+    QgsPointLocator_VisitorCentroidsInRect( QgsPointLocator *pl, QgsPointLocator::MatchList &lst, const QgsRectangle &srcRect, QgsPointLocator::MatchFilter *filter = nullptr )
+      : mLocator( pl )
+      , mList( lst )
+      , mSrcRect( srcRect )
+      , mFilter( filter )
+    {}
+
+    void visitNode( const INode &n ) override { Q_UNUSED( n ); }
+    void visitData( std::vector<const IData *> &v ) override { Q_UNUSED( v ); }
+
+    void visitData( const IData &d ) override
+    {
+      QgsFeatureId id = d.getIdentifier();
+      const QgsGeometry *geom = mLocator->mGeoms.value( id );
+      const QgsPointXY centroid = geom->centroid().asPoint();
+      if ( mSrcRect.contains( centroid ) )
+      {
+        QgsPointLocator::Match m( QgsPointLocator::Centroid, mLocator->mLayer, id, 0, centroid, -1 );
+
+        // in range queries the filter may reject some matches
+        if ( !( mFilter && !mFilter->acceptMatch( m ) ) )
+          mList << m;
+      }
+    }
+
+  private:
+    QgsPointLocator *mLocator = nullptr;
+    QgsPointLocator::MatchList &mList;
+    QgsRectangle mSrcRect;
+    QgsPointLocator::MatchFilter *mFilter = nullptr;
+};
+
+/**
+ * \ingroup core
+ * Helper class used when traversing the index looking for middle segment - builds a list of matches.
+ * \note not available in Python bindings
+ * \since QGIS 3.10
+*/
+class QgsPointLocator_VisitorMiddlesInRect : public IVisitor
+{
+  public:
+    //! Constructs the visitor
+    QgsPointLocator_VisitorMiddlesInRect( QgsPointLocator *pl, QgsPointLocator::MatchList &lst, const QgsRectangle &srcRect, QgsPointLocator::MatchFilter *filter = nullptr )
+      : mLocator( pl )
+      , mList( lst )
+      , mSrcRect( srcRect )
+      , mFilter( filter )
+    {}
+
+    void visitNode( const INode &n ) override { Q_UNUSED( n ); }
+    void visitData( std::vector<const IData *> &v ) override { Q_UNUSED( v ); }
+
+    void visitData( const IData &d ) override
+    {
+      QgsFeatureId id = d.getIdentifier();
+      const QgsGeometry *geom = mLocator->mGeoms.value( id );
+
+      QgsAbstractGeometry::vertex_iterator it = geom->vertices_begin();
+      QgsAbstractGeometry::vertex_iterator itPrevious = geom->vertices_begin();
+      it++;
+      for ( ; it != geom->vertices_end(); ++it, ++itPrevious )
```

Comment:
It does not seem correct to use vertex iterator here - in case of polygons with holes or multi-part geometries, this would also pretend that midpoints are in places where they do not exist...

Reply:
Right!
Done. Thanks